### PR TITLE
NPM cache: migrate ddb to s3 

### PR DESCRIPTION
**Pull request by edgarpoce**
**Goal**
migrate npm cache registry entry from ddb to s3

**Implementation**
- removed ddb table
- now the same bucket is used to store both the registry entry and the version tarballs. index is stored under the library folder with index.gz as the file name. There shouldn't be any collision because tarballs are stored with .tgz as the extension.
- refactored lambdas in the scope of this work with a more object oriented design
- added unit tests
- added seamless migration support. when autoupdate is executed the registry entry is stored in s3.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -3,3 +3,4 @@
 /coverage
 docs
 node_modules
+.vscode
```
Comment (troyready) [inline .gitignore]: Very minor thing here, but can you add an .editorconfig file in this PR than enforces a LF at end of files and add one here?

File: package.json [hunk 2]
```diff
@@ -25,11 +25,13 @@
     "@types/jest": "^24.0.20",
     "@types/node": "^11.10.4",
     "@types/request-promise-native": "^1.0.17",
+    "@types/sinon": "^9.0.0",
     "jest": "^24.9.0",
     "prettier": "^1.19.1",
     "serverless": "^1.38.0",
     "serverless-iam-roles-per-function": "^2.0.2",
     "serverless-webpack": "^5.2.0",
+    "sinon": "^9.0.2",
```
Comment (troyready) [inline package.json]: If at all possible, I'd like to stick with jest mocking instead of bringing in another framework.

I found some notes on a transition [here](https://github.com/skovhus/jest-codemods/issues/68#issuecomment-363054214).

File: src/npm/NpmPackage.test.ts [hunk 3]
```diff
@@ -0,0 +1,146 @@
+import * as AWS from "aws-sdk";
+import { NpmPackage } from "./NpmPackage";
+import * as sinon from "sinon";
+import * as zlib from "zlib";
```
Comment (troyready) [inline src/npm/NpmPackage.test.ts]: I could be missing something obviously, but I don't think the zlib import is needed anymore

File: src/npm/NpmPackage.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
+import * as request from "request-promise-native";
+import { S3Cache } from "./S3Cache";
+import { version } from "punycode";
+
+export class NpmPackage {
```
Comment (troyready) [inline src/npm/NpmPackage.ts:7]: As much as possible, it'd be nice to have jsdoc comments for each module/class/function, for our own future reference and so we're making the best possible typedoc (`npm run docs`) pages.

File: src/npm/autoupdate/autoupdate.ts [hunk 5]
```diff
@@ -3,98 +3,25 @@
...
 /** AWS Lambda entrypoint */
 export let handler: Handler = async (
   event: ScheduledEvent,
   context: Context,
 ): Promise<any> => {
-  const paginator = new ScanPaginator((DDBDocClient as unknown) as DynamoDB, {
-    TableName: tableName,
-  });
-
+  const s3Cache = new S3Cache();
```
Comment (troyready) [inline src/npm/autoupdate/autoupdate.ts]: We should aim to instantiate as many objects as possible outside of the handler, so they get reused on subsequent invocations.

The most notable ones being the AWS client objects, like the S3 client in S3Cache here.

**Review by troyready — CHANGES_REQUESTED** [review]
Great start. Added my nitpicky thoughts inline.

File: src/npm/NpmPackage.test.ts (hunk 3 above)
Comment (edgarpoce) [reply src/npm/NpmPackage.test.ts]: fixed in https://github.com/troyready/serverless-artifact-cacher/pull/3/commits/70c39de8aebb36bd15f6ae719496e8da52180cc5

File: src/npm/autoupdate/autoupdate.ts (hunk 5 above)
Comment (edgarpoce) [reply src/npm/autoupdate/autoupdate.ts]: are you OK if we add an IOC (inversion of control) lib?. e.g. http://inversify.io/
https://en.wikipedia.org/wiki/Inversion_of_control

File: .gitignore (hunk 1 above)
Comment (edgarpoce) [reply .gitignore]: fixed in https://github.com/troyready/serverless-artifact-cacher/pull/3/commits/557332fa22b21d6a481fd5f9f5b3185f71137357

File: src/npm/autoupdate/autoupdate.ts (hunk 5 above)
Comment (troyready) [reply src/npm/autoupdate/autoupdate.ts]: My preference would be against it (the focus in this project is simplicity & speed), but I'm not wholly opposed to it. Would need to be [sailplane injector](https://github.com/onicagroup/sailplane/tree/master/injector) (or bottlejs directly) to match our organizational preference/experience.

File: src/npm/NpmPackage.ts (hunk 4 above)
Comment (edgarpoce) [reply src/npm/NpmPackage.ts:7]: I'll change this. I didn't know it was the convention for this project.  
As a side note I'd like to mention that in most cases my personal preference is to avoid unnecessary documentation when module/class/function have descriptive names and the functionality is focused and matches the name of the containing block of code.

Comment (troyready) [reply src/npm/NpmPackage.ts:7]: Totally reasonable, feel free to use your best judgement, but assume you work with simpletons (me) that will have a hard time inferring functionality from anything non-obvious

File: src/npm/autoupdate/autoupdate.ts (hunk 5 above)
Comment (edgarpoce) [reply src/npm/autoupdate/autoupdate.ts]: Agreed, I'll create all instances in a place where they can be reused and accessed from the handler. Thanks!

File: package.json (hunk 2 above)
Comment (edgarpoce) [reply package.json]: makes sense. Thanks for pointing this out. I thought jest was limited to test execution and didn't know it contained mock functionality. Thanks!

File: src/npm/autoupdate/autoupdate.ts (hunk 5 above)
Comment (edgarpoce) [reply src/npm/autoupdate/autoupdate.ts]: fixed in https://github.com/troyready/serverless-artifact-cacher/pull/3/commits/bd79d9d2e98691c0219e903884666ad37f570966

File: src/npm/NpmPackage.ts (hunk 4 above)
Comment (edgarpoce) [reply src/npm/NpmPackage.ts:7]: fixed in https://github.com/troyready/serverless-artifact-cacher/pull/3/commits/bd79d9d2e98691c0219e903884666ad37f570966

File: package.json (hunk 2 above)
Comment (edgarpoce) [reply package.json]: fixed in https://github.com/troyready/serverless-artifact-cacher/pull/3/commits/ee0cb6bfd24076a9b7c5ad6fcf4c7998d32825d6.

Thanks for the link with the instructions. it was very helpful. thanks!

File: .editorconfig [hunk 6]
```diff
@@ -0,0 +1 @@
+end_of_line = lf
```
Comment (troyready) [inline .editorconfig]: Let's also add `insert_final_newline` (and then update .editorconfig & .gitignore) to have them

File: src/npm/NpmPackage.ts [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    });
+    return this.cache.put(this.npmPackageName, JSON.stringify(registryEntry));
+  }
+
+  /**
+   * Get or create DDB entry with map of files for package and return it
+   *
+   * Registry data is stored compressed in DDB for 2 reasons:
+   * 1) Not all registry data maps cleanly to a DDB item (e.g. empty strings)
+   * 2) Some items (e.g. @types/node) have too much data to store uncompressed
```
Comment (troyready) [inline src/npm/NpmPackage.ts]: I believe the details of this comment are out of date now.

File: jest.config.js [hunk 8]
```diff
@@ -1,10 +1,10 @@
 module.exports = {
   "coverageThreshold": {
     "global": {
-      "branches": 80,
+      "branches": 70,
       "functions": 80,
       "lines": 80,
-      "statements": -10
+      "statements": 80
```
Comment (troyready) [inline jest.config.js]: I realize the settings here are fairly arbitrary (and the existing ones were just the default values), but is it feasible to get back to the previous default thresholds?

I'd be open to doing it in a subsequent PR if it would be easier to manage after getting this PR through, or we can keep working in it.

File: src/npm/package/package.ts [hunk 9]
```diff
@@ -10,80 +10,18 @@ import {
...
-    "https://" +
-    event!.requestContext.domainName +
-    "/" +
-    event!.requestContext.stage +
-    "/npm-dlredirect/" +
-    npmPackageName +
-    "/";
-
+  const npmPackageName = decodeURIComponent(event!.pathParameters!.proxy);
+  const npmPackage = app.getNpmPackage(npmPackageName);
```
Comment (troyready) [inline src/npm/package/package.ts]: I realize now that some of the coverage issues here are totally my fault -- I thought this project already had more coverage, specifically including the handlers, but I realize now I was getting it confused with another project (e.g. [handler](https://github.com/troyready/git-lfs-s3/blob/master/src/batch/batch.ts#L135-L180) / [test](https://github.com/troyready/git-lfs-s3/blob/master/src/batch/batch.test.ts)).

So it's not a new problem now, but I have some concern about the gap in testing around the handler and NpmApp class itselff. I'm open to your suggestions on a path forward (I'm a little too narrow focused at the moment on catching every code path so I'd like to focus more testing at the entry point), but as I said in another comment that doesn't need to be in this PR.

**Review by troyready — CHANGES_REQUESTED** [review]

**Comment by troyready** [comment]
@edgarpoce I neglected to include this in my last review as well: some of the top-level README needs to be updated to reflect these changes

File: jest.config.js (hunk 8 above)
Comment (edgarpoce) [reply jest.config.js]: sure, I'll implement the tests as requested by the original threshold as it was before I added the test cases

I changed it because I added a high coverage on the modified code and the threshold required more. The missing cases that jest coverage is demanding don't seem to match any real use case scenario. e.g. receiving undefined as a response in one of the response attributes from s3.

File: .editorconfig (hunk 6 above)
Comment (edgarpoce) [reply .editorconfig]: fixed in https://github.com/troyready/serverless-artifact-cacher/pull/3/commits/4fb64dbf687dfe60dbcfb3ffbaaeb02e2633d413

File: src/npm/NpmPackage.ts (hunk 7 above)
Comment (edgarpoce) [reply src/npm/NpmPackage.ts]: fixed in https://github.com/troyready/serverless-artifact-cacher/pull/3/commits/acdf4fa2003d9db2bb534ef109ec1582574927cc

File: src/npm/package/package.ts (hunk 9 above)
Comment (edgarpoce) [reply src/npm/package/package.ts]: I refactored and added tests for the handlers involved in the scope of this PR.
fixed in https://github.com/troyready/serverless-artifact-cacher/pull/3/commits/672162b433d3a7f766e7909e125a20489cf9822a

File: jest.config.js (hunk 8 above)
Comment (edgarpoce) [reply jest.config.js]: fixed in https://github.com/troyready/serverless-artifact-cacher/pull/3/commits/9cc8b81b58b0c1ec17b56718d66e27b02418d39e.
it was not necessary to implement more tests because the relative weight of these edge cases now don't affect the threshold.

**Comment by edgarpoce** [comment]
> @edgarpoce I neglected to include this in my last review as well: some of the top-level README needs to be updated to reflect these changes

fixed in https://github.com/troyready/serverless-artifact-cacher/pull/3/commits/d5e4c96de69ec10696f926642649b297869ccde3

**Review by troyready — APPROVED** [review]
